source/adapters/hip/enqueue.cpp

```diff
@@ -770,29 +770,68 @@ UR_APIEXPORT ur_result_t UR_APICALL urEnqueueMemBufferCopyRect(
 ur_result_t commonMemSetLargePattern(hipStream_t Stream, uint32_t PatternSize,
                                      size_t Size, const void *pPattern,
                                      hipDeviceptr_t Ptr) {
-  // Calculate the number of patterns, stride, number of times the pattern
-  // needs to be applied, and the number of times the first 32 bit pattern
-  // needs to be applied.
-  auto NumberOfSteps = PatternSize / sizeof(uint8_t);
-  auto Pitch = NumberOfSteps * sizeof(uint8_t);
-  auto Height = Size / NumberOfSteps;
-  auto Count32 = Size / sizeof(uint32_t);
 
   // Get 4-byte chunk of the pattern and call hipMemsetD32Async
+  auto Count32 = Size / sizeof(uint32_t);
   auto Value = *(static_cast<const uint32_t *>(pPattern));
   UR_CHECK_ERROR(hipMemsetD32Async(Ptr, Value, Count32, Stream));
-  for (auto step = 4u; step < NumberOfSteps; ++step) {
-    // take 1 byte of the pattern
-    Value = *(static_cast<const uint8_t *>(pPattern) + step);
 
-    // offset the pointer to the part of the buffer we want to write to
-    auto OffsetPtr = reinterpret_cast<void *>(reinterpret_cast<uint8_t *>(Ptr) +
-                                              (step * sizeof(uint8_t)));
+  auto memsetRemainPattern = [&Stream, &pPattern,
+                              &Ptr](const auto Size, const auto PatternSize) {
+    // Calculate the number of patterns, stride and the number of times the
+    // pattern needs to be applied.
+    auto NumberOfSteps = PatternSize / sizeof(uint8_t);
+    auto Pitch = NumberOfSteps * sizeof(uint8_t);
+    auto Height = Size / NumberOfSteps;
 
-    // set all of the pattern chunks
-    UR_CHECK_ERROR(hipMemset2DAsync(OffsetPtr, Pitch, Value, sizeof(uint8_t),
-                                    Height, Stream));
-  }
+    for (auto step = 4u; step < NumberOfSteps; ++step) {
+      // take 1 byte of the pattern
+      auto Value = *(static_cast<const uint8_t *>(pPattern) + step);
+
+      // offset the pointer to the part of the buffer we want to write to
+      auto OffsetPtr = reinterpret_cast<void *>(
+          reinterpret_cast<uint8_t *>(Ptr) + (step * sizeof(uint8_t)));
+
+      // set all of the pattern chunks
+      UR_CHECK_ERROR(hipMemset2DAsync(OffsetPtr, Pitch, Value, sizeof(uint8_t),
+                                      Height, Stream));
+    }
+  };
+  // There is a bug in ROCm prior to 6.0.0 version which causes hipMemset2D and
+  // hipMemset3D to behave incorrectly when acting on host pinned memory. In
+  // such a case the following part of memsetting the remaining part of the
+  // pattern is emulated with memcpy.
+#if HIP_VERSION < 60000000
+  hipPointerAttribute_t ptrAttribs{};
+  UR_CHECK_ERROR(hipPointerGetAttributes(&ptrAttribs, (const void *)Ptr));
+
+  const bool ptrIsHost{ptrAttribs.memoryType == hipMemoryTypeHost};
```

Comment:
Since ROCm 5.7.1 the `memoryType`  for shared allocations will override `hipMemoryTypeHost` and `hipMemoryTypeDevice` with `hipMemoryTypeManaged`, so it's not ideal to rely on this way. I see you have checked explicitly `!ptrAttribs.isManaged` below, which is sufficient to tell you what  you need in this case, but I just wanted to let you know to be aware of those changes.
I think another possible way to check for a _HIP host allocation_ is `ptrAttribs.hostPointer && !ptrAttribs.isManaged`, but as long as `!isManaged` then your check will work anyways, so not blocking the PR.blocking

Reply:
Aha yes, I changed it to `hostPointer && !isManaged`, thanks for the suggestion!

Reply:
That working as expected? I never tested this suggestion myself.

Reply:
Yeah that's the right behaviour. Nice, good to know.